### PR TITLE
perf(engine): segment-scope SDR→HDR preflight

### PR DESCRIPTION
**Pull request by jrusso1020**
## What

Scopes the SDR→HDR preflight re-encode to the segment the composition actually uses, mirroring the existing VFR→CFR segment-scope fix.

## Why

`convertSdrToHdr` was re-encoding entire source files, so a 30-minute SDR screen recording contributing a 2-second clip in a mixed HDR/SDR composition ate multi-second preflight time that produced frames no one would ever read. Validated on a mixed 30s-SDR + 2s-HDR fixture: `hdrPreflightMs` drops **87%** (1162→148ms), `videoExtractMs` drops **82%** (1272→231ms), `tmpPeakBytes` drops **45%** (8.2MB→4.5MB).

Depends on #444 (phase-level instrumentation) for the measurement surface.

## How

- `convertSdrToHdr` gains `startTime` and `duration` parameters ahead of the upstream `targetTransfer` arg added by #370. New signature: `convertSdrToHdr(input, output, startTime, duration, targetTransfer, signal, config)`. `-ss $start -t $duration` is added to the ffmpeg args.
- Phase 2 now captures the full `VideoMetadata` per `resolvedVideos` entry (previously just `colorSpace`) so the caller can compute `segDuration` from `video.end - video.start` with a fallback to `metadata.durationSeconds - video.mediaStart` for unbounded (Infinity) clips — without firing another ffprobe.
- After a successful convert, `entry.video.mediaStart` is zeroed out via shallow-copy (doesn't mutate the caller's `VideoElement`) so downstream extraction seeks from 0 instead of the original offset. Mirrors what the VFR→CFR path already does.

## Test plan

Validation on `/tmp/hf-fixtures/hdr-sdr-mixed-scope`:
```
hdrPreflightMs: >1000 → 150   (gate: <300)   ✓
videoExtractMs: 1272  → 237   (-82%)
tmpPeakBytes:   8.2MB → 4.5MB (-45%)
```

- [x] Unit test: new regression test synthesizes 10s SDR + 2s HDR fixture inline and asserts the converted file's duration matches the 2s used segment (pre-fix matched the 10s source)
- [x] Lint + format
- [x] Typecheck
- [x] Manual perf validation against synthesized fixture

### CODE REVIEW

**Comment by jrusso1020** [comment]
* **#454** <a href="https://app.graphite.com/github/pr/heygen-com/hyperframes/454?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#446** <a href="https://app.graphite.com/github/pr/heygen-com/hyperframes/446?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#445** <a href="https://app.graphite.com/github/pr/heygen-com/hyperframes/445?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/heygen-com/hyperframes/445?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#444** <a href="https://app.graphite.com/github/pr/heygen-com/hyperframes/444?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by miguel-heygen — COMMENTED** [review]
## Review: perf(engine): segment-scope SDR→HDR preflight

Correct approach — mirrors the established VFR segment-scope pattern. `-ss` before `-i` (input seeking), `-t` after `-i`, smart metadata reuse avoids extra ffprobe, shallow copy prevents caller mutation. The claimed perf numbers (87% drop in `hdrPreflightMs`) are consistent with scoping a 30-min transcode to 2 seconds.

### P1 — Important (inherited, not a regression)

**`mediaStart` beyond source duration → silent empty output.** If `entry.video.mediaStart > metadata.durationSeconds`, `-ss` seeks past EOF and FFmpeg silently produces a 0-byte file. The subsequent `entry.videoPath = convertedPath` then points to a broken file. Same edge case exists in the VFR path. Since this code is being touched, worth adding:
```ts
if (startTime >= metadata.durationSeconds) { errors.push(...); continue; }
```

### P2 — Should fix

**No validation that `duration > 0` before passing to FFmpeg.** If `end === start`, `segDuration = 0`, FFmpeg receives `-t 0`, and silently produces an empty output. One-line guard:
```ts
if (duration <= 0) {
  throw new Error(`SDR→HDR conversion: duration must be positive (got ${duration})`);
}
```

### P3 — Suggestions

- **Test lives inside the "on a VFR source" describe block** but is about HDR preflight scoping. A separate `describe` block (or renaming to something broader like "extractAllVideoFrames preflight normalization") would improve readability.
- **Test uses `extractVideoMetadata`** while production code uses `extractMediaMetadata`. If the latter is canonical going forward, the test should match.
- **Benchmark reproducibility** — the PR body references `/tmp/hf-fixtures/hdr-sdr-mixed-scope` which isn't committed. Consider adding a script reference for anyone wanting to reproduce the numbers.

### Verdict: Approve with minor suggestions

Core change is correct and well-tested. Push for the `duration > 0` guard — it's a one-line fix that prevents a silent failure mode.

**Comment by jrusso1020** [comment]
Rebased onto the updated #444 head and pushed the P2 + P1 fixes from <https://github.com/heygen-com/hyperframes/pull/445#pullrequestreview-4164954688|your review>:

1. *P2 — `duration <= 0` throws inside `convertSdrToHdr`.* One-line guard at the top of the function, so every caller (current + future) is protected without having to re-derive the invariant at the call site. Message format: `convertSdrToHdr: duration must be positive (got ${duration})`.
2. *P1 — `mediaStart >= durationSeconds` skip-and-error at the HDR call site.* Using `errors.push({...}) + continue` (your suggested shape), not a throw — keeps the rest of the render going for other entries. The error text names both values so the ops side can see which source overshot.

Scope note on P1: the same edge case exists in the VFR path too as you pointed out, but I kept the fix tight to HDR since that's what this PR actually touches. Happy to mirror it into the VFR path in a follow-up so the symmetry holds and we don't paper over the inherited issue forever.

On the P3s:
- *Move the regression test out of the "VFR source" describe block* — agree it reads oddly there. Easy rename in a follow-up; leaving as-is here to keep the diff focused on the semantic fixes.
- *`extractVideoMetadata` vs `extractMediaMetadata`* — agree the test should match production. Same follow-up PR.
- *Benchmark fixture not committed* — noted; the `/tmp/hf-fixtures/hdr-sdr-mixed-scope` path was local. If we want the 87% / 82% / 45% numbers reproducible by reviewers, a `scripts/bench/hdr-scope-fixture.sh` script that builds the mixed fixture on demand would do it. Deferring to a perf-tooling follow-up.

Engine tests still green (459/459) with the new guards in place.

File: packages/engine/src/services/videoFrameExtractor.ts [hunk 1]
```diff
@@ -483,11 +497,45 @@ export async function extractAllVideoFrames(
         // SDR video in a mixed timeline — convert to the dominant HDR transfer
         // so the encoder tags the final video correctly (PQ vs HLG).
         const entry = resolvedVideos[i];
-        if (!entry) continue;
+        const metadata = videoMetadata[i];
+        if (!entry || !metadata) continue;
+
+        // Guard against mediaStart past EOF — FFmpeg's `-ss` silently produces
+        // a 0-byte file when seeking beyond the source duration, and the
```
Comment (miguel-heygen) [inline packages/engine/src/services/videoFrameExtractor.ts:511]: P3: This guard only skips HDR normalization; it does not skip later extraction. The invalid entry stays in `resolvedVideos`, so Phase 3 still calls `extractVideoFramesRange()` with the same past-EOF `mediaStart` and emits a second raw FFmpeg error for the same clip. I reproduced this on the current head with a synthetic 2s SDR + 1s HDR case and `mediaStart=5`, which returned both the intended descriptive skip error and a downstream `Nothing was written into output file` failure. If the intent is a true skip, this entry needs to be removed or marked non-extractable before Phase 3.

Comment (jrusso1020) [reply packages/engine/src/services/videoFrameExtractor.ts:511]: Great catch — `errors.push({...}); continue;` only short-circuited the normalization step, not Phase 3 extraction. Fixed in `17d1622`:

- Track HDR-preflight-skipped indices in a `Set<number>` (hoisted outside the `if (hdrInfo.hasHdr...)` block).
- After the HDR preflight loop completes and `breakdown.hdrPreflightMs` is recorded, iterate backwards through `resolvedVideos` and `.splice(i, 1)` from every parallel array (`resolvedVideos`, `videoMetadata`, `videoColorSpaces`) at each skipped index.
- Phase 2b (VFR preflight) and Phase 3 (extract) now only see validated entries, so the past-EOF clip gets exactly one descriptive error instead of the duplicate `Nothing was written into output file` failure you reproduced.

Downstack #446 introduces a 4th parallel array (`cacheKeyInputs` built pre-Phase-2) — extended the same splice to that array in a follow-up commit (`a0f49f2`) on the #446 branch so the cache lookup at `cacheKeyInputs[i]` doesn't misalign.

Left the VFR path untouched on this PR since you originally flagged that edge case as "inherited, not a regression" — splicing symmetry for VFR is a natural follow-up once this lands.

480/480 engine tests pass.

**Review by miguel-heygen — APPROVED** [review]
Re-checked the live head after the past-EOF follow-up. The new splice removes skipped HDR entries before later phases, my synthetic past-EOF repro now yields exactly one descriptive error, and the latest regression CI is green.

**Review by vanceingalls — APPROVED** [review]

**Comment by jrusso1020** [comment]
### Merge activity

* **Apr 24, 4:15 AM UTC**: @jrusso1020 merged this pull request with [Graphite](https://app.graphite.com/github/pr/heygen-com/hyperframes/445).
